src/responder/pam/pamsrv_gssapi.c

```diff
@@ -0,0 +1,593 @@
+/*
+    Authors:
+        Pavel Březina <pbrezina@redhat.com>
+
+    Copyright (C) 2020 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <errno.h>
+#include <gssapi.h>
+#include <gssapi/gssapi_ext.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <talloc.h>
+
+#include "responder/common/responder_packet.h"
+#include "responder/common/responder.h"
+#include "responder/common/cache_req/cache_req.h"
+#include "responder/pam/pamsrv.h"
+#include "sss_client/sss_cli.h"
+#include "util/util.h"
+#include "util/sss_utf8.h"
+
+static errno_t read_str(size_t body_len,
+                        uint8_t *body,
+                        size_t *pctr,
+                        const char **_str)
+{
+    size_t i;
+
+    for (i = *pctr; i < body_len && body[i] != 0; i++) {
+        /* counting */
+    }
+
+    if (i >= body_len) {
+        return EINVAL;
+    }
+
+    if (!sss_utf8_check(&body[*pctr], i - *pctr)) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Body is not UTF-8 string!\n");
+        return EINVAL;
+    }
+
+    *_str = (const char *)&body[*pctr];
+    *pctr = i + 1;
+
+    return EOK;
+}
+
+static bool pam_gssapi_allowed(struct pam_ctx *pam_ctx,
+                               struct sss_domain_info *domain,
+                               const char *service)
+{
+    char **list = pam_ctx->gssapi_services;
+
+    if (domain->gssapi_services != NULL) {
```

Comment:
Hi,

what about sub-domains? I guess if there is no sub-domain specific setting it would be reasonable to check the parent first before falling back to the pam responder setting?

bye,
Sumit

Reply:
Do you thing it would be useful to allow inheritance of this option and its explicit configuration per subdomain?

Reply:
Hi,

I was more thinking about what would be expected. E.g. if you join your client to an AD domain which is part of a forest and set `pam_gssapi_services` in the `[domain/...]` section my expectation would be that it works for users from other domains in the forest as well.

With inheritance enabled it would be a bit odd to disable the feature for a sub-domain because you have to use `pam_gssapi_services = NotExistingServiceName` since iirc we cannot handle empty values correctly.

But currently if you want to allow sub-domain user you have to add `pam_gssapi_services` to the `[pam]` section which enables it globally. So if you prefer to avoid inheritance I'm fine with only sub-domain specific configuration, but it should be explained in the man page that domain-specific settings must be set for each domain, including sub-domains, individually.

bye,
Sumit